src/main/scala/net/snowflake/spark/snowflake/LoggerWrapper.scala

```diff
@@ -0,0 +1,166 @@
+    */
+  def trace(msg: String): Unit = {
+    logger.trace(msg)
+    sendLogTelemetryIfEnabled(LoggerWrapper.TRACE_LEVEL, msg)
+  }
+
+  /**
+    * Log a message at the DEBUG level.
+    *
+    * @param msg the message string to be logged
+    */
+  def debug(msg: String): Unit = {
+    logger.debug(msg)
+    sendLogTelemetryIfEnabled(LoggerWrapper.DEBUG_LEVEL, msg)
+  }
+
+  /**
+    * Log a message at the INFO level.
+    *
+    * @param msg the message string to be logged
+    */
+  def info(msg: String): Unit = {
+    logger.info(msg)
+    sendLogTelemetryIfEnabled(LoggerWrapper.INFO_LEVEL, msg)
+  }
+
+  /**
+    * Log a message at the WARN level.
+    *
+    * @param msg the message string to be logged
+    */
+  def warn(msg: String): Unit = {
+    logger.warn(msg)
+    sendLogTelemetryIfEnabled(LoggerWrapper.WARN_LEVEL, msg)
+  }
+
+  /**
+    * Log a message at the WARN level.
+    *
+    * @param msg the message string to be logged
+    * @param t   the exception (throwable) to log
+    */
+  def warn(msg: String, t: Throwable): Unit = {
+    logger.warn(msg, t)
+    sendLogTelemetryIfEnabled(LoggerWrapper.WARN_LEVEL, s"$msg: ${t.getMessage}")
+  }
+
+  /**
+    * Log a message at the ERROR level.
+    *
+    * @param msg the message string to be logged
+    */
+  def error(msg: String): Unit = {
+    logger.error(msg)
+    sendLogTelemetryIfEnabled(LoggerWrapper.ERROR_LEVEL, msg)
+  }
+
+  /**
+    * Log an exception (throwable) at the ERROR level with an
+    * accompanying message.
+    *
+    * @param msg the message accompanying the exception
+    * @param t   the exception (throwable) to log
+    */
+  def error(msg: String, t: Throwable): Unit = {
+    logger.error(msg, t)
+    sendLogTelemetryIfEnabled(LoggerWrapper.ERROR_LEVEL, s"$msg: ${t.getMessage}")
+  }
+
+  /**
+    * Log a message at the ERROR level according to the specified format
+    * and arguments.
+    * <p/>
+    * <p>This form avoids superfluous object creation when the logger
+    * is disabled for the ERROR level. </p>
+    *
+    * @param format the format string
+    * @param arg1   the first argument
+    * @param arg2   the second argument
+    */
+  def error(format: String, arg1: Any, arg2: Any): Unit = {
+    logger.error(format, arg1, arg2)
+    val logEntry = try {
+      // Never throw exception
+      MessageFormatter.format(format, arg1, arg2).getMessage
+    } catch {
+      case _: Throwable => {
+        s"$format, ${arg1.toString}, ${arg2.toString}"
+      }
+    }
+    sendLogTelemetryIfEnabled(LoggerWrapper.ERROR_LEVEL, logEntry)
+  }
+}
+
+private[snowflake] object LoggerWrapper {
+  // Log levels
+  val TRACE_LEVEL = "TRACE"
+  val DEBUG_LEVEL = "DEBUG"
+  val INFO_LEVEL = "INFO"
+  val WARN_LEVEL = "WARN"
+  val ERROR_LEVEL = "ERROR"
+}
+
+/**
+  * The <code>LoggerWrapperFactory</code> is a utility class producing a
+  * <code>LoggerWrapper</code> object.
+  */
+private[snowflake] object LoggerWrapperFactory {
```

Comment:
Do you really need this object ? Can't you just call
new LoggerWithTelemetry(logger) wherever needed ?

Reply:
update it.